Utilities/Mergers/src/MergerAlgorithm.cxx

```diff
@@ -0,0 +1,109 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file Merger.cxx
+/// \brief Implementation of O2 Mergers, v0.1
+///
+/// \author Piotr Konopka, piotr.jan.konopka@cern.ch
+
+#include "Mergers/MergerAlgorithm.h"
+
+#include "Mergers/MergeInterface.h"
+
+#include <TH1.h>
+#include <TH2.h>
+#include <TH3.h>
+#include <THn.h>
+#include <TTree.h>
+#include <THnSparse.h>
+#include <TObjArray.h>
+
+namespace o2::mergers::algorithm
+{
+
+void merge(TObject* const target, TObject* const other)
+{
+  if (target == nullptr) {
+    throw std::runtime_error("Merging target is nullptr");
+  }
+  if (other == nullptr) {
+    throw std::runtime_error("Object to be merged in is nullptr");
+  }
+  if (other == target) {
+    throw std::runtime_error("Merging target and the other object point to the same address");
+  }
+  // fixme: should we check if names match?
+
+  // We expect that both objects follow the same structure, but we allow to add missing objects to TCollections.
+
+  if (auto custom = dynamic_cast<MergeInterface*>(target)) {
+
+    custom->merge(dynamic_cast<MergeInterface* const>(other));
+
+  } else if (auto targetCollection = dynamic_cast<TCollection*>(target)) {
+
+    // We assume that the other has the same structure
+    auto otherCollection = reinterpret_cast<TCollection*>(other);
+
+    auto otherIterator = otherCollection->MakeIterator();
+    while (auto otherObject = otherIterator->Next()) {
+      TObject* targetObject = targetCollection->FindObject(otherObject->GetName());
+      if (targetObject) {
+        // That might be another collection or a concrete object to be merged, we walk on the collection recursively.
+        merge(targetObject, otherObject);
+      } else {
+        // We prefer to clone instead of passing the pointer in order to simplify deleting the `other`.
+        targetCollection->Add(otherObject->Clone());
+      }
+    }
+
+  } else {
+    Long64_t errorCode = 0;
+    TObjArray otherCollection;
+    otherCollection.SetOwner(false);
+    otherCollection.Add(other);
+
+    const char* className = target->ClassName();
+    if (strncmp(className, "TH1", 3) == 0) {
```

Comment:
Avoid slow string comparison
```suggestion
    if (target->Class() == TH1::Class()) {
```

Reply:
That doesn't actually work, probably because `target->Class()` is TObject there. Also, even comparing TH1 and TH1F would return false. I can rely on `InheritsFrom(const TClass *cl) const` instead, but it is actually 1.5x slower than comparing strings. We could reduce the code to this though:
```
    if (target->InheritsFrom(TH1::Class())) {
      // this includes TH1, TH2, TH3
      errorCode = reinterpret_cast<TH1*>(target)->Merge(&otherCollection);
    } else if (target->InheritsFrom(THnBase::Class())) {
      // this includes THn and THnSparse
      errorCode = reinterpret_cast<THnBase*>(target)->Merge(&otherCollection);
    } else if (target->InheritsFrom(TTree::Class())) {
      errorCode = reinterpret_cast<TTree*>(target)->Merge(&otherCollection);
    } 
```
Which makes it probably even...